### PR TITLE
Remove agent MaxNumWorkers hard limit

### PR DESCRIPTION
**Pull request by AlexandreYang**
### What does this PR do?

Remove max check workers hard limit related to [`check_runners` config](https://github.com/DataDog/datadog-agent/blob/eb0b08db9191c9ee526dd42bb0c72fa573e31e46/pkg/config/config_template.yaml#L380-L390).

### Motivation

It seems the limit is here because most integrations are in python, and the agent won't be able to handle 25+ python instances being run concurrently. But 25+ go corecheck instances being run concurrently is not an issue.

Use case: Users might use integrations like corecheck SNMP integration with thousands of check instances.
Currently, the limit is 25 check runners and that's not enough to run thousands of check instances with reasonable collection interval (e.g. 15-60 sec).

For example, with 25 runners and expected 15 sec collection interval, with 1sec per check run (snmp check might take multiple seconds to complete), we can only run at most ~375 (25 * 15) instances. 

If the customer have more than ~375 instances, that will lead to lower collection interval e.g. 1000 instances with 25 check runners will lead to ~40 sec collection interval.

A possible reason it hasn't been an issue previously is that we didn't have a use case with such large # of check instances 1000+ that is "heavy" to run (`snmp` is "heavier" to run than `ping` or `http_check` integration).

### Is there any risk to remove the hard limit ?

In theory removing the hard limit might cause issue to users if they have a config with a high # of check runners `check_runners: 100` and in the past the agent was limited to 25 check runners, but with the removal of the hard limit, they have now more check runners than before. The new higher number of check runners might create bottle neck elsewhere:
- Aggregator/Forwarder level (if there are a lot more metrics being processed by the agent)
- Too many Python check instances being run in parallel

That being said, I think this theoretical case described above might be rare, an edge case.

If that's an issue, see Alternative.

### Alternative solution

An alternative solution is to add a new option like `disable_check_runners_limit`, set to `false` by default.

### Describe your test plan

- Test we can use more than 25 check runners.
- Test with a high number of SNMP instances e.g. 1000 instances

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -1260,11 +1260,6 @@ func setNumWorkers(config Config) {
 		numWorkers = 1
 	}
 
-	if numWorkers > MaxNumWorkers {
```
Comment (hush-hush) [inline pkg/config/config.go:1263]: Do we still need the `MaxNumWorkers` variable ?

Comment (hush-hush) [reply pkg/config/config.go:1263]: We might still want to print a warning or something telling users that this will have impact, so it's not enable in the wrong case

Comment (AlexandreYang) [reply pkg/config/config.go:1263]: > Do we still need the `MaxNumWorkers` variable ?

Yes, [here](https://github.com/DataDog/datadog-agent/blob/e081bed84866686eeab56a98da5ff9b4b9f03ded/pkg/collector/runner/runner.go#L111-L124). I think it's good to keep the limit 25 for this logic setting # of workers based on the # of check instances.

> We might still want to print a warning or something telling users that this will have impact, so it's not enable in the wrong case

Agree with having a clear message. It seems warning level might a bit too high since it will be a normal use case to set a # of check runner >25 if this PR is merged.

I can add a `info` for >25 runners to warn people, e.g. >25 might not work well if they are running many Python checks

We currently already log here the # of check runners here: https://github.com/DataDog/datadog-agent/blob/e081bed84866686eeab56a98da5ff9b4b9f03ded/pkg/collector/runner/runner.go#L92

**Comment by olivielpeau** [comment]
Thanks for the detailed description!

When `check_runners` is set to a non-zero value I think it's fine for the limit of `25` not to apply. The edge case that you describe of users that have set a high value for `check_runners` is good to keep in mind, but at worst in this case this change will cause a small performance degradation that will be simple to address (solution: set `check_runners` to `25`).

The limit of 25 only makes sense when `check_runners` is set to `0` which means that the number of check runners is dynamic. This has never been the default, so I doubt many users are using this, but keeping the limit in this case is safer. We may want to remove it later if it also becomes a limitation in the future, but then we'll have to consider the performance consequences more closely.

File: releasenotes/notes/agent_remove_max_wokers_hard_limit-41fd2c1873d89549.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+enhancements:
+  - |
+    Remove agent max workers hard limit
```
Comment (hush-hush) [inline releasenotes/notes/agent_remove_max_wokers_hard_limit-41fd2c1873d89549.yaml:4]: We have many types of workers in the agent, I think we should be more precise about which ones we're talking about.

Also could we explain in which case a users might want to increase this past 25.

File: pkg/config/config.go (hunk 1 above)
Comment (hush-hush) [reply pkg/config/config.go:1263]: Thanks for the answer. I think it's fine then. We might want to add a warning later if we have too many support cases about python performances with a high number of workers.

File: releasenotes/notes/agent_remove_max_wokers_hard_limit-41fd2c1873d89549.yaml (hunk 2 above)
Comment (AlexandreYang) [reply releasenotes/notes/agent_remove_max_wokers_hard_limit-41fd2c1873d89549.yaml]: Sure improved the release note here: https://github.com/DataDog/datadog-agent/pull/7507/commits/7f4501824c64346600b5f91ecfacf8ab3f7b64a7

**Review by hush-hush — APPROVED** [review]
